no-restricted-imports doesn't restrict export * when importNames are configured

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are here for another reason, please see below:

    1. To propose a new rule: https://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a rule change: https://eslint.org/docs/developer-guide/contributing/rule-changes
    3. To request a change that is not a bug fix, rule change, or new rule: https://eslint.org/docs/developer-guide/contributing/changes
    4. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

<!--
    If you are using ESLint v6.5.0 or later, you can run ESLint with the `--env-info` flag and paste the output here.
-->

* **ESLint Version:** v6.8.0
* **Node Version:** v12.14.0
* **npm Version:** v6.13.4

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```js
module.exports = {
    parserOptions: {
        ecmaVersion: 2015,
        sourceType: "module"
    },
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

[Online Demo Link](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtaW1wb3J0czogW1wiZXJyb3JcIixcbiAge1xuICAgIG5hbWU6IFwibW9kMVwiICBcbiAgfSxcbiAge1xuICAgIG5hbWU6IFwibW9kMlwiLFxuICAgIGltcG9ydE5hbWVzOiBbXCJhXCJdXG4gIH1cbl0qL1xuXG5pbXBvcnQgKiBhcyBmb28gZnJvbSBcIm1vZDFcIjsgLy8gZXJyb3JcbmV4cG9ydCAqIGZyb20gXCJtb2QxXCI7IC8vIGVycm9yXG5cbmltcG9ydCB7IGEgfSBmcm9tIFwibW9kMlwiOyAvLyBlcnJvclxuZXhwb3J0IHsgYSB9IGZyb20gXCJtb2QyXCI7IC8vIGVycm9yXG5pbXBvcnQgKiBhcyBiYXIgZnJvbSBcIm1vZDJcIjsgLy8gZXJyb3JcbmV4cG9ydCAqIGZyb20gXCJtb2QyXCI7IC8vIG5vIGVycm9yXG5cbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

<!-- Paste the source code below: -->
```js
/* eslint no-restricted-imports: ["error",
  {
    name: "mod1"  
  },
  {
    name: "mod2",
    importNames: ["a"]
  }
]*/

import * as foo from "mod1"; // error
export * from "mod1"; // error

import { a } from "mod2"; // error
export { a } from "mod2"; // error
import * as bar from "mod2"; // error
export * from "mod2"; // no error, is this a bug?
```

<!-- Paste the command you used to run ESLint: -->
```bash
eslint index.js
```

**What did you expect to happen?**

Errors for the first 5 declarations.

I'm not sure what should be expected behavior for `export * from "mod2"`. That is the question of this issue.

**What actually happened? Please include the actual, raw output from ESLint.**

Errors for the first 5 declarations, as expected.

No error for `export * from "mod2"`.

```
  11:1  error  'mod1' import is restricted from being used                no-restricted-imports
  12:1  error  'mod1' import is restricted from being used                no-restricted-imports
  14:1  error  'mod2' import is restricted from being used                no-restricted-imports
  15:1  error  'mod2' import is restricted from being used                no-restricted-imports
  16:1  error  * import is invalid because 'a' from 'mod2' is restricted  no-restricted-imports
```

**Are you willing to submit a pull request to fix this bug?**

Yes.

--- a/lib/rules/no-restricted-imports.js
+++ b/lib/rules/no-restricted-imports.js
@@ -91,6 +91,7 @@ module.exports = {
     },
 
     create(context) {
+        const sourceCode = context.getSourceCode();
         const options = Array.isArray(context.options) ? context.options : [];
         const isPathAndPatternsObject =
             typeof options[0] === "object" &&
@@ -216,30 +217,36 @@ module.exports = {
          */
         function checkNode(node) {
             const importSource = node.source.value.trim();
-            const importNames = node.specifiers ? node.specifiers.reduce((map, specifier) => {
-                let name;
-                const specifierData = { loc: specifier.loc };
+            const importNames = new Map();
 
-                if (specifier.type === "ImportDefaultSpecifier") {
-                    name = "default";
-                } else if (specifier.type === "ImportNamespaceSpecifier") {
-                    name = "*";
-                } else if (specifier.imported) {
-                    name = specifier.imported.name;
-                } else if (specifier.local) {
-                    name = specifier.local.name;
-                }
+            if (node.type === "ExportAllDeclaration") {
+                const starToken = sourceCode.getFirstToken(node, 1);
 
-                if (name) {
-                    if (map.has(name)) {
-                        map.get(name).push(specifierData);
-                    } else {
-                        map.set(name, [specifierData]);
+                importNames.set("*", [{ loc: starToken.loc }]);
+            } else if (node.specifiers) {
+                for (const specifier of node.specifiers) {
+                    let name;
+                    const specifierData = { loc: specifier.loc };
+
+                    if (specifier.type === "ImportDefaultSpecifier") {
+                        name = "default";
+                    } else if (specifier.type === "ImportNamespaceSpecifier") {
+                        name = "*";
+                    } else if (specifier.imported) {
+                        name = specifier.imported.name;
+                    } else if (specifier.local) {
+                        name = specifier.local.name;
+                    }
+
+                    if (name) {
+                        if (importNames.has(name)) {
+                            importNames.get(name).push(specifierData);
+                        } else {
+                            importNames.set(name, [specifierData]);
+                        }
                     }
                 }
-
-                return map;
-            }, new Map()) : new Map();
+            }
 
             checkRestrictedPathAndReport(importSource, importNames, node);
 

--- a/tests/lib/rules/no-restricted-imports.js
+++ b/tests/lib/rules/no-restricted-imports.js
@@ -168,6 +168,17 @@ ruleTester.run("no-restricted-imports", rule, {
         {
             code: "import {\nAllowedObject,\nDisallowedObject, // eslint-disable-line\n} from \"foo\";",
             options: [{ paths: [{ name: "foo", importNames: ["DisallowedObject"] }] }]
+        },
+        {
+            code: "export * from \"foo\";",
+            options: ["bar"]
+        },
+        {
+            code: "export * from \"foo\";",
+            options: [{
+                name: "bar",
+                importNames: ["DisallowedObject"]
+            }]
         }
     ],
     invalid: [{
@@ -354,6 +365,37 @@ ruleTester.run("no-restricted-imports", rule, {
             line: 1,
             column: 8,
             endColumn: 16
+        }]
+    },
+    {
+        code: "export * from \"foo\";",
+        options: [{
+            paths: [{
+                name: "foo",
+                importNames: ["DisallowedObject"],
+                message: "Please import 'DisallowedObject' from /bar/ instead."
+            }]
+        }],
+        errors: [{
+            message: "* import is invalid because 'DisallowedObject' from 'foo' is restricted. Please import 'DisallowedObject' from /bar/ instead.",
+            type: "ExportAllDeclaration",
+            line: 1,
+            column: 8,
+            endColumn: 9
+        }]
+    },
+    {
+        code: "export * from \"foo\";",
+        options: [{
+            name: "foo",
+            importNames: ["DisallowedObject1, DisallowedObject2"]
+        }],
+        errors: [{
+            message: "* import is invalid because 'DisallowedObject1, DisallowedObject2' from 'foo' is restricted.",
+            type: "ExportAllDeclaration",
+            line: 1,
+            column: 8,
+            endColumn: 9
         }]
     },
     {

